### PR TITLE
4.5 - ORM\DeleteQuery shims

### PR DESCRIPTION
**Pull request by markstory**
Add a forwards compatibility implementation of the upcoming ORM\Query changes. My hope is that through deprecations and this new runtime interface enforcement solution will allow users to incrementally upgrade what I think could be an otherwise harsh break in compatibility.

This will create some merge conflicts with 5.x that I'm happy to take ownership of solving. If folks don't hate this approach I'll continue for all the other query types.

I've considered doing the same for the Database package to help our small but still important standalone package user base.

### CODE REVIEW

File: src/ORM/Exception/DeprecationException.php [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+namespace Cake\ORM\Exception;
+
+use Cake\Core\Exception\CakeException;
+
+class DeprecationException extends CakeException
+{
+    /**
+     * @inheritDoc
+     */
+    protected $_messageTemplate = 'Deprecated: %s';
```
Comment (othercorey) [inline src/ORM/Exception/DeprecationException.php]: The 5.x deprecation message includes the version number. Can we do it here too?

File: src/ORM/Query/DeleteQuery.php [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+        parent::__construct($connection, $table);
+        $this->from([$table->getAlias() => $table->getTable()]);
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function delete(?string $table = null)
+    {
+        $this->_deprecatedException('delete()', 'Remove this method call.');
```
Comment (othercorey) [inline src/ORM/Query/DeleteQuery.php]: This isn't deprecated if we're throwing an exception. Somehow we need to communicate this call is not needed with the new DeleteQuery object.

Comment (markstory) [reply src/ORM/Query/DeleteQuery.php]: Should we be using regular deprecation warningsninstead of exceptions for along these methods then?

I was trying to make the required updates easy to find. However, using an exception will give a tougher upgrade than a warning would.

I will update these to use a warning instead.

**Comment by markstory** [comment]
Last call on this. I'd like to keep making progress on the other query types.

**Comment by othercorey** [comment]
Looks ok

**Comment by ADmad** [comment]
Looks good, just the phpcs/stan errors need to be addressed.

**Comment by othercorey** [comment]
patch has 7.1%, are you going to instantiate these shims somewhere?

**Comment by markstory** [comment]
> are you going to instantiate these shims somewhere?

I've been trying to think of how to reasonably test the shims as well. Delete queries don't have all of these methods and wouldn't use the query parts if they were set. I could write tests that convert DeleteQuery into the other query types and ensure that DeleteQuery can still be coerced into an insert/update/select query.

**Comment by markstory** [comment]
@othercorey I added some rudimentary tests for commonly used methods and use cases. Are there others you'd like to see? I'm planning on having each deprecation emitting class go through the same tests via a provider.

**Comment by othercorey** [comment]
> @othercorey I added some rudimentary tests for commonly used methods and use cases. Are there others you'd like to see? I'm planning on having each deprecation emitting class go through the same tests via a provider.

It's fine with me as long as people can instantiate the classes without fatal errors.

**Pull request merged**